app/src/main/java/org/mozilla/fenix/tabstray/syncedtabs/SyncedTabs.kt

```diff
@@ -0,0 +1,339 @@
+import androidx.compose.ui.text.font.Font
+import androidx.compose.ui.text.font.FontFamily
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.unit.sp
+import androidx.navigation.NavController
+import mozilla.components.browser.storage.sync.SyncedDeviceTabs
+import mozilla.components.browser.storage.sync.TabEntry
+import mozilla.components.browser.toolbar.MAX_URI_LENGTH
+import mozilla.components.feature.syncedtabs.view.SyncedTabsView
+import org.mozilla.fenix.NavGraphDirections
+import org.mozilla.fenix.R
+import org.mozilla.fenix.components.components
+import org.mozilla.fenix.compose.PrimaryText
+import org.mozilla.fenix.compose.SecondaryText
+import org.mozilla.fenix.ext.toShortUrl
+import org.mozilla.fenix.theme.FirefoxTheme
+import mozilla.components.browser.storage.sync.Tab as SyncTab
+
+/**
+ * Top-level list UI for displaying Synced Tabs in the Tabs Tray.
+ *
+ * @param syncedTabs [SyncedTabsListItem] The tab UI items to be displayed.
+ * @param listener [SyncedTabsView.Listener] to apply to this layout node.
+ */
+@Composable
+fun SyncedTabsList(syncedTabs: List<SyncedTabsListItem>, listener: SyncedTabsView.Listener) {
+    val listState = rememberLazyListState()
+    LazyColumn(
+        state = listState,
+        verticalArrangement = Arrangement.spacedBy(8.dp),
+        modifier = Modifier.fillMaxSize()
+    ) {
+        items(syncedTabs) { syncedTabItem ->
+            when (syncedTabItem) {
+                is SyncedTabsListItem.Device -> SyncedTabsDeviceItem(syncedTabItem.displayName)
+                is SyncedTabsListItem.Error -> SyncedTabsErrorItem(
+                    syncedTabItem.errorType,
+                    syncedTabItem.navController
+                )
+                is SyncedTabsListItem.NoTabs -> SyncedTabsNoTabsItem()
+                is SyncedTabsListItem.Tab -> SyncedTabsTabItem(
+                    syncedTabItem.tab.active().title,
+                    syncedTabItem.tab.active().url.toShortUrl(components.publicSuffixList).take(MAX_URI_LENGTH),
+                ) { listener.onTabClicked(syncedTabItem.tab) }
+            }
+        }
+    }
+}
+
+/**
+ * Text header for sections of synced tabs
+ *
+ * @param deviceName The name of the user's device connected that has synced tabs.
+ */
+@Composable
+fun SyncedTabsDeviceItem(deviceName: String) {
+    PrimaryText(
+        text = deviceName,
+        fontSize = 12.sp,
+        fontFamily = FontFamily(Font(R.font.metropolis_semibold)),
+        letterSpacing = 0.04.sp,
+        maxLines = 1,
+        modifier = Modifier
+            .padding(16.dp, 16.dp, 8.dp, 7.dp)
+            .fillMaxWidth()
+    )
+}
+
+/**
+ * Synced tab list item UI
+ *
+ * @param tabTitleText The tab's display text.
+ * @param url The tab's URL.
+ * @param onClick The click handler when this synced tab is clicked.
+ */
+@Composable
+fun SyncedTabsTabItem(tabTitleText: String, url: String, onClick: () -> Unit) {
```

Comment:
titles can be empty strings, in which case we show the full url (until MAX_URI_LENGTH). I don't see that case handled.

Reply:
That logic wasn't in the now-deleted SyncedTabsViewHolder. I'm happy to add this in though

Reply:
Please do. It was there at one point, and it might have been removed.

Reply:
Update: it looks like this is the behavior for inactive tabs but it never existed in synced tabs. I'll add it here now.